### PR TITLE
Allow multiple listeners on same port in Gateway API provider

### LINKED ISSUE #9026: Allow multiple listeners on same port in Gateway API provider
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

Concern K8SGatewayAPI provider: 
Sould be possible to have multiple HTTP/HTTPS listeners on the same port (the goal of a reverse proxy :)
Now this is forbidden by this line : https://github.com/traefik/traefik/blob/1048348ae65497feb4a05e090f0ac910586c7996/pkg/provider/kubernetes/gateway/kubernetes.go#L343

Am I right ?

### PR DESCRIPTION
**Pull request by burner-account**
### What does this PR do?

The Gateway API spec allows for multiple listeners with the same (protocol,port)-tuple on gateways.
This PR removes a limitation that was preventing this behavior.

Currently, this multi-listener support is only active for HTTPS protocol, but can easily be extended to other protocols with a minor modification of the shareableListenerProtocols slice (see listener.go).

### Motivation

closes #9026 

### More

- [x] Added/updated tests

listener_test.go contains some test that allows (if extended) to verify that the behavior matches the spec.

### Additional Notes

make test actually failed with:
```
--- FAIL: TestLookupAddress (0.00s)
    --- FAIL: TestLookupAddress/IP_doesn't_need_refresh (130.51s)
        proxy_test.go:207: 
                Error Trace:    proxy_test.go:207
                Error:          Received unexpected error:
                                dial tcp 8.8.4.4:53: connect: connection timed out
                Test:           TestLookupAddress/IP_doesn't_need_refresh
    --- FAIL: TestLookupAddress/Hostname_needs_refresh (261.58s)
        proxy_test.go:207: 
                Error Trace:    proxy_test.go:207
                Error:          Received unexpected error:
                                dial tcp 8.8.8.8:53: connect: connection timed out
                Test:           TestLookupAddress/Hostname_needs_refresh

```
These DNS servers/ports are simply blocked by my infrastructure. The Gateway API tests are run before these failing tests.

make pull-images ✔
make validate ✔
make test ✔

### CODE REVIEW

**Comment by burner-account** [comment]
> At glance, regarding the check of the uniqueness of each listener (the combination of Hostname, Port, and Protocol) for a Gateway, I am wondering why not create a map key composed of the hostname, port, and protocol, to check if it has already been used?

Because that would allow you to create nonsensical stuff like:
```
(443, HTTPS, "hostname.com") -> "443-https-hostname.com"
(443, HTTP, "hostname.com") -> "443-http-hostname.com"

--> no key collision? good to go!
```
And: From my point of view, the underlying technical limitations (bound port) should be reflected in the code. Therefore i chose to have block of listeners grouped by ports.

**Comment by burner-account** [comment]
Or phrased differently:

To simply have unique 3-tuple (port, hostname, protocol) is a very relaxed requirement that satisfies the Gateway API spec.

That you cannot have different protocols on the same port is a limitation unrelated to the Gateway API spec. But having only one protocol per port is unavoidable.

**Comment by burner-account** [comment]
> Traefik has no limitation handling UDP/TCP/HTTP/HTTPS on the same port, so it seems accurate to use a unique 3-tuple (port, hostname, protocol). Can you address the rework in that way?

Hi @rtribotte! This is news - but interesting news - to me. I have to admit, i did not look up how traefik handles sockets and ran with an assumption.

I will rework the PR to reflect this. ETA 16h from now.

**Comment by burner-account** [comment]
Ok. I need help. 

I followed your advice on traefik supporting different protocols on one port at the same time.
The test cases for my code pass. The logic itself is quite simple.

```
--- FAIL: TestLoadMixedRoutes/Mixed_routes_with_All_namespace_selector
--- FAIL: TestLoadMixedRoutes/Mixed_routes_with_core_group
--- FAIL: TestLoadMixedRoutes/Mixed_routes

... and many more
```
Did I do that?

**Comment by rtribotte** [comment]
Hello @burner-account,

Thanks for addressing those changes.
I will push some adjustments, I removed the "shareable protocols" check because it doesn’t seem either that we needed to enforce that. It is also what was breaking the unit tests.

**Review by rtribotte — APPROVED** [review]
Thanks 👍

File: pkg/provider/kubernetes/gateway/fixtures/mixed/with_multiple_listeners_using_same_hostname_port_protocol.yml [hunk 1]
```diff
@@ -54,7 +75,17 @@ spec:
         namespaces:
           from: Same
 
-    - name: tls
+---
+kind: Gateway
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: my-gateway-http
```
Comment (kevinpollet) [inline pkg/provider/kubernetes/gateway/fixtures/mixed/with_multiple_listeners_using_same_hostname_port_protocol.yml:82]:
```suggestion
  name: my-gateway-tls
```

**Review by jbdoumenjou — APPROVED** [review]
:+1:

**Pull request merged**